### PR TITLE
Ensure that path parameters are not empty strings

### PR DESCRIPTION
**Pull request by douglasmiller**
Adding a validation function to `InterpolatePath` to ensure that path parameters are not empty strings.

Adds a test to ensure that the path parameters are properly encoded.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -94,16 +94,33 @@ func newClient(apiKey string, httpClient *http.Client) *Client {
...
 
+func validatePathParameters(params []string) error {
+	invalidParams := []string{}
+	for _, param := range params {
+		if len(strings.TrimSpace(param)) == 0 {
+			invalidParams = append(invalidParams, param)
+		}
+	}
+	if len(invalidParams) > 0 {
+		return fmt.Errorf("Parameters cannot be empty strings")
```
Comment (bhelx) [inline client.go:105]: As of right now, we don't return any generic errors. I'm wondering what the best practice is. in the examples, we don't handle the generic case.

Comment (douglasmiller) [reply client.go:105]: Agreed, I wasn't sure what recurly error would be appropriate here. I'm open to thoughts.

Comment (douglasmiller) [reply client.go:105]: I suppose that the error Class would be `ErrorClassClient`, but would it make more sense for the error Type to be `ErrorTypeUnknown`, `ErrorTypeBadRequest`, or something else?

Comment (bhelx) [reply client.go:105]: I don't know. there is no "blowing up" like there is in other languages. maybe we just change the examples.

Comment (bhelx) [reply client.go:105]: Seems there are actually some other places we do this. so regardless, we will need to consider updating the examples.

File: client_test.go [hunk 2]
```diff
@@ -5,6 +5,49 @@ import (
...
+	scenario := &Scenario{
+		T: t,
+		AssertRequest: func(req *http.Request) {
+			t.Assert(req.Method, http.MethodGet, "HTTP Method")
+			t.Assert(req.URL.String(), "https://v3.recurly.com/resources/abcd1234", "Request URL")
+			// assert headers and other request properties
+		},
+		MakeResponse: func(req *http.Request) *http.Response {
+			// default headers set, we may want to customize though
+			return mockResponse(req, 200, `{"id": "abcd1234"}`)
```
Comment (bhelx) [inline client_test.go:42]: I changed the interface of this to be a pointer to a string. Please change to: 
```
String(`{"id": "abcd1234"}`)
```

File: client_test.go [hunk 3]
```diff
@@ -5,6 +5,49 @@ import (
...
+	scenario := &Scenario{
+		T: t,
+		AssertRequest: func(req *http.Request) {
+			t.Assert(req.Method, http.MethodGet, "HTTP Method")
+			t.Assert(req.URL.String(), "https://v3.recurly.com/resources/%2F", "Request URL")
+			// assert headers and other request properties
+		},
+		MakeResponse: func(req *http.Request) *http.Response {
+			// default headers set, we may want to customize though
+			return mockResponse(req, 200, `{"id": "abcd1234"}`)
```
Comment (bhelx) [inline client_test.go:20]: Same here, the body should be a pointer

File: client.go (hunk 1 above)
Comment (douglasmiller) [reply client.go:105]: Nearly every example has to be updated to account for the potential error being returned from the validation anyway, so it's not that much more of a lift

File: client.go [hunk 4]
```diff
@@ -95,16 +95,37 @@ func newClient(apiKey string, httpClient *http.Client) *Client {
...
+	for _, param := range params {
+		if len(strings.TrimSpace(param)) == 0 {
+			invalidParams = append(invalidParams, param)
+		}
+	}
+	if len(invalidParams) > 0 {
+		return &Error{
+			Message: "Operation parameters cannot be empty strings",
+			Class:   ErrorClassClient,
+			Type:    ErrorTypeValidation,
```
Comment (bhelx) [inline client.go:109]: I think the problem with making this a validation error, is these are mostly shown to the customer. I'd rather this be a BadRequest error or something that isn't mean to be caught.

Comment (bhelx) [reply client.go:109]: ideally this hits the part of their error handling that just logs it and kills the request with a generic error.

Comment (douglasmiller) [reply client.go:109]: Agreed, I used Validation just to continue the discussion with a `recurly.Error` being used. I actually debated defining a new `ErrorTypeX` to be used in cases like this that aren't actual API response based errors

**Pull request merged**